src/Storages/MergeTree/MergeTreeBaseSelectProcessor.cpp

```diff
@@ -71,6 +77,87 @@ MergeTreeBaseSelectProcessor::MergeTreeBaseSelectProcessor(
 }
 
 
+bool MergeTreeBaseSelectProcessor::getNewTask()
+{
+    auto get_from_buffer = [&]()
+    {
+        /// Try to get from buffer
+        while (!buffered_ranges.empty())
+        {
+            auto ranges = std::move(buffered_ranges.front());
+            buffered_ranges.pop_front();
+
+            assert(!ranges.empty());
+
+            if (Status::Accepted == performRequestToCoordinator(ranges))
+                return true;
+        }
+        return false;
+    };
+
+    auto get_delayed = [&]()
+    {
+        /// Try to get from buffer
+        while (!delayed_tasks.empty())
+        {
+            task = std::move(delayed_tasks.front());
+            delayed_tasks.pop_front();
+
+            assert(!task->mark_ranges.empty());
+
+            if (Status::Accepted == performRequestToCoordinator(task->mark_ranges))
+                return true;
+        }
+
+        finish();
+        return false;
+    };
+
+    if (get_from_buffer())
+        return true;
+
+    if (no_more_tasks)
+        return get_delayed();
+
+    while (true)
+    {
+        auto res = getNewTaskImpl();
+
+        /// It means that parallel_reading feature is disabled
+        if (!extension.has_value())
+        {
+            if (res)
+                finalizeNewTask();
+
+            return res;
+        }
+
+        /// The end of execution. No task.
+        if (!res)
+        {
+            no_more_tasks = true;
+            return get_delayed();
+        }
+
+        if (task->mark_ranges.empty())
+            continue;
+
+        auto predicate = [this](MarkRange range)
+        {
+            auto what = fmt::format("{}_{}_{}", task->data_part->info.getPartName(), range.begin, range.end);
```

Comment:
Here, hash is actually a part of protocol. If you change it a little bit in a new version, you will get a cache miss.
I suppose we should make a hash calculation a part of `PartitionReadRequest`. Maybe even add version for it.

Reply:
Maybe, instead of converting to string, we can hash one-by-one using prev value as a seed for `CityHash64WithSeed`.

Reply:
Also I have a fear that for different replicas we can get different sets of mark ranges (it is possible if replicas have different versions wit a little bit different pk analysis). Don't know what to do in this case.

Reply:
We can check the server revision of replicas on the initiator and disable parallel reading in case they are different. It will be quite tricky and will require an additional roundtrip.